network/gossip/libp2p/peerManager_test.go

```diff
@@ -114,108 +115,128 @@ func (ts *PeerManagerTestSuite) TestUpdatePeers() {
+
+	// makes a request for peer update
+	wg.Add(1) // expects a call to `ConnectPeers` by requesting peer update
 	pm.RequestPeerUpdate()
 
 	// assert that a call to connect to peers is made
-	assert.Eventually(ts.T(), func() bool {
-		return connector.AssertNumberOfCalls(ts.T(), "ConnectPeers", 2)
-	}, 10*time.Millisecond, 1*time.Millisecond)
+	unittest.RequireReturnsBefore(suite.T(), wg.Wait, 1*time.Millisecond,
+		"ConnectPeers is not running on request")
 }
 
 // TestConcurrentOnDemandPeerUpdate tests that concurrent on-demand peer update request never block
-func (ts *PeerManagerTestSuite) TestConcurrentOnDemandPeerUpdate() {
+func (suite *PeerManagerTestSuite) TestConcurrentOnDemandPeerUpdate() {
 	currentIDs := unittest.IdentityListFixture(10)
 	idProvider := func() (flow.IdentityList, error) {
 		return currentIDs, nil
 	}
 
-	ctx, cancel := context.WithCancel(ts.ctx)
+	ctx, cancel := context.WithCancel(suite.ctx)
 	defer cancel()
 
 	connector := new(mock.Connector)
 	// connectPeerGate channel gates the return of the connector
 	connectPeerGate := make(chan time.Time)
 	defer close(connectPeerGate)
-	connector.On("ConnectPeers", ctx, testifymock.Anything).Return(nil).WaitUntil(connectPeerGate)
-	connector.On("DisconnectPeers", ctx, testifymock.Anything).Return(nil)
 
-	pm := NewPeerManager(ctx, ts.log, idProvider, connector)
+	// creates mock connector
+	wg := &sync.WaitGroup{} // keeps track of number of calls on `ConnectPeers`
+	// we expect it to be called twice, i.e.,
+	// one for periodic update and one for the on-demand request
+	count, times := 0, 2
+	wg.Add(times)
+	connector.On("ConnectPeers", ctx, testifymock.Anything).Run(func(args testifymock.Arguments) {
+		if count < times {
+			count++
+			wg.Done()
+		}
+	}).Return(nil).
+		WaitUntil(connectPeerGate) // blocks call for connectPeerGate channel
+	connector.On("DisconnectPeers", ctx, testifymock.Anything).Return(nil)
+	pm := NewPeerManager(ctx, suite.log, idProvider, connector)
 
 	// set the periodic interval to a high value so that periodic runs don't interfere with this test
 	PeerUpdateInterval = time.Hour
 
 	// start the peer manager
 	// this should trigger the first update and which will block on the ConnectPeers to return
 	err := pm.Start()
-	assert.NoError(ts.T(), err)
-
-	// make 10 concurrent request for peer update
-	wg := sync.WaitGroup{}
-	for i := 0; i < 10; i++ {
-		wg.Add(1)
-		go func() {
-			pm.RequestPeerUpdate()
-			wg.Done()
-		}()
-	}
+	assert.NoError(suite.T(), err)
 
-	// assert that none of the request is blocked even if update is blocked
-	unittest.AssertReturnsBefore(ts.T(), wg.Wait, time.Second)
+	// makes 10 concurrent request for peer update
+	unittest.RequireConcurrentCallsReturnBefore(suite.T(), pm.RequestPeerUpdate, 10, time.Second,
+		"concurrent peer update requests could not return on time")
 
-	// allow the first update to finish
+	// allow the first and second updates to finish
+	connectPeerGate <- time.Now()
```

Comment:
An extra return could also mean that we have either:
`periodic run`, `first on demand run` OR `periodic run`, `first on demand`, `second on demand`.
the latter should never happen. Hence, I would just let everyone block and then just open the gate once and count the calls.

Reply:
right, fixed in https://github.com/onflow/flow-go/pull/99/commits/ae48d421a30f79378446acabcdc2483c30e9f264.